client/ip_addr.go

```diff
@@ -14,112 +13,74 @@ type IpAddress struct {
-	c, err := client.getMikrotikClient()
+var _ Resource = (*IpAddress)(nil)
 
-	if err != nil {
-		return nil, err
-	}
-
-	cmd := Marshal("/ip/address/add", addr)
-	log.Printf("[INFO] Running the mikrotik command: `%s`", cmd)
-	r, err := c.RunArgs(cmd)
-
-	log.Printf("[DEBUG] ip address creation response: `%v`", r)
-
-	if err != nil {
-		return nil, err
-	}
-
-	id := r.Done.Map["ret"]
-
-	return client.FindIpAddress(id)
+func (b *IpAddress) ActionToCommand(a Action) string {
+	return map[Action]string{
+		Add:    "/ip/address/add",
+		Find:   "/ip/address/print",
+		Update: "/ip/address/set",
+		Delete: "/ip/address/remove",
+	}[a]
 }
 
-func (client Mikrotik) ListIpAddress() ([]IpAddress, error) {
-	c, err := client.getMikrotikClient()
-
-	if err != nil {
-		return nil, err
-	}
-	cmd := []string{"/ip/address/print"}
-	log.Printf("[INFO] Running the mikrotik command: `%s`", cmd)
-	r, err := c.RunArgs(cmd)
-
-	if err != nil {
-		return nil, err
-	}
-	log.Printf("[DEBUG] found ip address: %v", r)
-
-	ipaddr := []IpAddress{}
-
-	err = Unmarshal(*r, &ipaddr)
+func (b *IpAddress) IDField() string {
+	return ".id"
+}
 
-	if err != nil {
-		return nil, err
-	}
+func (b *IpAddress) ID() string {
+	return b.Id
+}
 
-	return ipaddr, nil
+func (b *IpAddress) SetID(id string) {
+	b.Id = id
 }
 
-func (client Mikrotik) FindIpAddress(id string) (*IpAddress, error) {
-	c, err := client.getMikrotikClient()
+func (b *IpAddress) AfterAddHook(r *routeros.Reply) {
+	b.Id = r.Done.Map["ret"]
+}
 
+// Typed wrappers
+func (c Mikrotik) AddIpAddress(r *IpAddress) (*IpAddress, error) {
+	res, err := c.Add(r)
 	if err != nil {
 		return nil, err
 	}
 
-	cmd := []string{"/ip/address/print", "?.id=" + id}
-	log.Printf("[INFO] Running the mikrotik command: `%s`", cmd)
-	r, err := c.RunArgs(cmd)
-
-	log.Printf("[DEBUG] ip address response: %v", r)
+	return res.(*IpAddress), nil
+}
 
+func (c Mikrotik) UpdateIpAddress(r *IpAddress) (*IpAddress, error) {
+	res, err := c.Update(r)
 	if err != nil {
 		return nil, err
 	}
 
-	ipaddr := IpAddress{}
-	err = Unmarshal(*r, &ipaddr)
+	return res.(*IpAddress), nil
+}
 
+func (c Mikrotik) FindIpAddress(id string) (*IpAddress, error) {
+	res, err := c.Find(&IpAddress{Id: id})
 	if err != nil {
 		return nil, err
 	}
 
-	if ipaddr.Id == "" {
-		return nil, NewNotFound(fmt.Sprintf("ip address `%s` not found", id))
-	}
-
-	return &ipaddr, nil
+	return res.(*IpAddress), nil
 }
 
-func (client Mikrotik) UpdateIpAddress(addr *IpAddress) (*IpAddress, error) {
-	c, err := client.getMikrotikClient()
-
+func (client Mikrotik) ListIpAddress() ([]IpAddress, error) {
+	res, err := client.List(&IpAddress{})
 	if err != nil {
 		return nil, err
 	}
-
-	cmd := Marshal("/ip/address/set", addr)
-	log.Printf("[INFO] Running the mikrotik command: `%s`", cmd)
-	_, err = c.RunArgs(cmd)
-
-	if err != nil {
-		return nil, err
+	returnSlice := make([]IpAddress, len(res))
+	for i, v := range res {
+		returnSlice[i] = *(v.(*IpAddress))
```

Comment:
Is it possible to type assert that this is a slice of crud client type (i.e. `[]IpAddress`) similar to how it works for the single result functions?

Regarding my comment above, we are creating 3 slices for each list of results. While I don't think this will cause performance issues for a terraform provider use case, it would be nice to simplify this if possible.

Reply:
no, it is not possible (or I am not aware of it) to assert slices.
I tried to do:
```go
	res, err := client.List(&IpAddress{})
	if err != nil {
		return nil, err
	}
	
	var i interface{} = res
	return i.([]IpAddress), nil
```
and got:
```shell
panic: interface conversion: interface {} is []client.Resource, not []client.IpAddress
```